### DISCUSSION TITLE
New volumetric fog implementation

### DISCUSSION THREAD

**Discussion by Try** [Try/OpenGothic#399] [Polls]
I'm testing new stochastic fog right now. While quality is definitely better, performance cost is quite noticeable.
Please test on you hardware how much slower this fog is and post fps before-vs-after on same scene.
Also fps with `Radial fog` disabled is interesting.

### UPD: Merged

![изображение](https://user-images.githubusercontent.com/122894/216782738-77743341-f5b2-4644-b31b-6216d564d5b7.png)
![изображение](https://user-images.githubusercontent.com/122894/216782741-53df8ae5-d2bf-4c3b-b10b-810a13f0928b.png)
![изображение](https://user-images.githubusercontent.com/122894/216782744-55646f6e-3ba4-4311-a2eb-063b8427c59f.png)

### Performance
Here is frame comparative chart. New rendering is 6.5% slower:
![изображение](https://user-images.githubusercontent.com/122894/216783033-aafd1aab-e19b-4996-a721-f90087ca1383.png)

Core part of algorithm is to use path tracing with `32` ray steps, + random offset for each pixel.
You may notice that `L1` cache (Yellow) is completely trashed, probably to large distance span between shadow-map samples, on each sample.
Tested with low-res (512x256) shadows - helps a little, but only a little.

### Testing

Here build to test with new shading:
https://ci.appveyor.com/project/Try/opengothic/builds/46140513/job/5j0a451cbalra790/artifacts

Please report, if it works well on your machine.

**Comment by Try** [comment]
UPD:
Want to encourage testings, before voting. 6.5% performance may not be representation to your GPU. In principle RTX measurements are hardly representative.

**Comment by thokkat** [comment]
Tested on a GTX 1060 in 2560x1440. Everything looked good so far.

Actually this would be the perfect scenario for an ingame benchmark to get comparable numbers.

Fps table:
|| scene 1| scene 2| scene 3| scene 4|scene 5|scene 6
 |-| -| -| -| -|-|-|
| **off**| 107| 122| 133| 139|139|129
| **new**| 103| 121| 133| 129|130|131
| **old**| 118| 136| 153| 161|159|145



<details>
  <summary> old/new fog screenshots</summary>

scene 1 old
![1o](https://user-images.githubusercontent.com/100468436/216792189-04fadbc0-fc20-4032-becb-b6b478d81127.png)

scene 1 new
![1n](https://user-images.githubusercontent.com/100468436/216792190-fbf37429-2668-4904-abff-3ba068b43599.png)

scene 2 old
![2o](https://user-images.githubusercontent.com/100468436/216792198-371e7b40-f283-488c-8b8b-6fe012d9d22e.png)

scene 2 new
![2n](https://user-images.githubusercontent.com/100468436/216792201-a9830b3c-da4a-4632-8505-e919ade3cf7f.png)

scene 3 old
![3o](https://user-images.githubusercontent.com/100468436/216792277-e6f6c717-bf90-4527-a698-9aa4f6b58375.png)

scene 3 new
![3n](https://user-images.githubusercontent.com/100468436/216792276-ec0f99f2-b250-405e-8dc8-17eb2fc0d53c.png)

scene 4 old
![4o](https://user-images.githubusercontent.com/100468436/216792286-83637aca-8ad7-4656-b94e-325e6e2ecc65.png)

scene 4 new
![4n](https://user-images.githubusercontent.com/100468436/216792287-a5f2b6a9-dbd7-4064-abda-41931352fe3b.png)

scene 5 old
![5o](https://user-images.githubusercontent.com/100468436/216792295-2eed4edc-15ef-46ec-a2df-1c33983e9eb7.png)

scene 5 new
![5n](https://user-images.githubusercontent.com/100468436/216792304-0efdab3e-1125-41fb-bbcf-a9e5694ca625.png)

scene 6 old
![6o](https://user-images.githubusercontent.com/100468436/216792326-219c0481-13a3-4759-9e1c-2a3206fd026b.png)

scene 6 new
![6n](https://user-images.githubusercontent.com/100468436/216792324-e5ad5a28-905d-4dab-a5e6-3bc27fe9f7f8.png)




</details>

**Reply by Try** [reply]
Hm, interesting that `off` is slower on your machine than `old`. Practically current `off` suppose to have no changes in compare to previous `off`.

**Reply by thokkat** [reply]
There used to be a 20 % performance penalty when activating fog. At some time `off` got lower fps, maybe when the sun got that large halo.

I compared with initial implementation (13th September build). Different scenes than in previous table due to save game incompatibility.

|| scene 1| scene 2| scene 3| 
 |-| -| -| -|
| **off initial**| 141| 135| 159
| **off current**| 120| 115| 126
| **new**| 123| 108| 127
| **old initial**| 120| 111| 129
| **old current**| 133| 128| 142

**Comment by Try** [comment]
After 3.5 long days of work:
![изображение](https://user-images.githubusercontent.com/122894/232342067-edf5ff02-b471-477f-b33a-e153dd9a6fa5.png)

0.80ms - occlusion lut
0.14ms - draw-pass
